src/internal/collection/collection_test.go

```diff
@@ -1,699 +1,1352 @@
+					for _, k := range modKeys {
+						err := writer(context.Background(), func(rw col.ReadWriteCollection) error {
+							testProto := &col.TestItem{}
+							if err := rw.Update(k, testProto, func() error {
+								testProto.Value = changedValue
+								return nil
+							}); err != nil {
+								return err
+							}
+							return nil
+						})
+						require.NoError(t, err)
+					}
+					testSort(t, modRead, col.SortByModRevision, modKeys)
+				})
+
+				subsuite.Run("Key", func(t *testing.T) {
+					testSort(t, defaultRead, col.SortByKey, idRange(0, 10))
+				})
+			})
+
+			subsuite.Run("ErrorInCallback", func(t *testing.T) {
+				t.Parallel()
+				count := 0
+				testProto := &col.TestItem{}
+				err := defaultRead.List(testProto, col.DefaultOptions(), func() error {
+					count++
+					return &TestError{}
+				})
+				require.YesError(t, err)
+				require.True(t, errors.Is(err, TestError{}), "Incorrect error: %v", err)
+				require.Equal(t, 1, count, "List callback was called multiple times despite erroring")
+			})
+
+			subsuite.Run("ErrBreak", func(t *testing.T) {
+				t.Parallel()
+				count := 0
+				testProto := &col.TestItem{}
+				err := defaultRead.List(testProto, col.DefaultOptions(), func() error {
+					count++
+					return errutil.ErrBreak
+				})
+				require.NoError(t, err)
+				require.Equal(t, 1, count, "List callback was called multiple times despite aborting")
+			})
+		})
+
+		suite.Run("Count", func(subsuite *testing.T) {
+			subsuite.Parallel()
+			subsuite.Run("Success", func(t *testing.T) {
+				t.Parallel()
+				count, err := defaultRead.Count()
+				require.NoError(t, err)
+				require.Equal(t, int64(10), count)
+
+				count, err = emptyRead.Count()
+				require.NoError(t, err)
+				require.Equal(t, int64(0), count)
+			})
+		})
+
+		// 'Watch' functions with callbacks don't have a good hook to trigger
+		// events. This is a bit racy, but we can sleep a bit to let the listen get
+		// started, then trigger some writes.
+		asyncWrite := func(t *testing.T, writer WriteCallback, cbs ...func(rw col.ReadWriteCollection) error) {
+			eg, ctx := errgroup.WithContext(context.Background())
+
+			eg.Go(func() error {
+				time.Sleep(100 * time.Millisecond)
+				for _, cb := range cbs {
+					if err := writer(ctx, cb); err != nil {
+						return err
+					}
+				}
+				return nil
+			})
+
+			t.Cleanup(func() {
+				require.NoError(t, eg.Wait())
+			})
 		}
-		if err := jobInfos.Get(j4.Job.ID, job); err != nil {
-			return err
+
+		// Helper function for checking watch events
+		collectEventsCallback := func(count int, out *[]*watch.Event) func(*watch.Event) error {
+			return func(ev *watch.Event) error {
+				*out = append(*out, ev)
+				if len(*out) == count {
+					return errutil.ErrBreak
+				}
+				return nil
+			}
 		}
 
-		jobInfos.Put(j1.Job.ID, j1)
-		if err := jobInfos.Get(j1.Job.ID, job); err != nil {
-			return err
+		collectEventsChannel := func(watcher watch.Watcher, count int, out *[]*watch.Event) {
```

Comment:
I think this should return a channel and then the callback is out of the picture and you can test by reading a certain number from the channel or timing out.

Reply:
Reworked these tests to get rid of `asyncWrite`, they are now in `watch_test.go` and I deduped the tests between the channel-based and callback-based API.